File: Updated_CMS-main/backend/routes/attendance.py
from flask import Blueprint, request, jsonify
from supabase_client import get_supabase
from datetime import datetime, timedelta
from functools import wraps

attendance_bp = Blueprint('attendance', __name__)

# Initialize Supabase client
supabase = get_supabase()

def handle_errors(f):
    @wraps(f)
    def wrapper(*args, **kwargs):
        try:
            return f(*args, **kwargs)
        except Exception as e:
            print(f"Error in {f.__name__}: {str(e)}")
            return jsonify({"success": False, "error": str(e)}), 500
    return wrapper

# Attendance Management
@attendance_bp.route('/attendance', methods=['GET'])
@handle_errors
def get_attendance():
    """Get attendance records with optional filtering"""
    query = supabase.table('attendance').select('''
        *,
        students (
            id,
            register_number,
            full_name,
            courses (
                name,
                code
            )
        ),
        subjects (
            id,
            name,
            code
        ),
        faculty (
            id,
            full_name,
            employee_id
        )
    ''')
    
    # Add filters if provided
    if request.args.get('student_id'):
        query = query.eq('student_id', request.args.get('student_id'))
    if request.args.get('subject_id'):
        query = query.eq('subject_id', request.args.get('subject_id'))
    if request.args.get('faculty_id'):
        query = query.eq('faculty_id', request.args.get('faculty_id'))
    if request.args.get('date'):
        query = query.eq('date', request.args.get('date'))
    if request.args.get('status'):
        query = query.eq('status', request.args.get('status'))
    if request.args.get('date_from') and request.args.get('date_to'):
        query = query.gte('date', request.args.get('date_from')).lte('date', request.args.get('date_to'))
    
    result = query.order('date', desc=True).execute()
    return jsonify({"success": True, "data": result.data})

@attendance_bp.route('/attendance', methods=['POST'])
@handle_errors
def mark_attendance():
    """Mark attendance for students"""
    data = request.get_json()
    
    # Support both single attendance and bulk attendance marking
    if 'attendance_records' in data:
        # Bulk attendance marking
        attendance_records = data['attendance_records']
        required_fields = ['student_id', 'subject_id', 'faculty_id', 'date', 'status']
        
        for record in attendance_records:
            for field in required_fields:
                if field not in record:
                    return jsonify({"success": False, "error": f"Missing required field: {field} in attendance record"}), 400
        
        # Validate all records first
        for record in attendance_records:
            # Check if student exists
            student_result = supabase.table('students').select('id').eq('id', record['student_id']).execute()
            if not student_result.data:
                return jsonify({"success": False, "error": f"Invalid student_id: {record['student_id']}"}), 400
            
            # Check if subject exists
            subject_result = supabase.table('subjects').select('id').eq('id', record['subject_id']).execute()
            if not subject_result.data:
                return jsonify({"success": False, "error": f"Invalid subject_id: {record['subject_id']}"}), 400
            
            # Check if faculty exists
            faculty_result = supabase.table('faculty').select('id').eq('id', record['faculty_id']).execute()
            if not faculty_result.data:
                return jsonify({"success": False, "error": f"Invalid faculty_id: {record['faculty_id']}"}), 400
            
            # Validate status
            valid_statuses = ['present', 'absent', 'late']
            if record['status'] not in valid_statuses:
                return jsonify({"success": False, "error": f"Invalid status: {record['status']}. Must be one of: {valid_statuses}"}), 400
        
        # Check for existing attendance records
        for record in attendance_records:
            existing = supabase.table('attendance').select('id').eq('student_id', record['student_id']).eq('subject_id', record['subject_id']).eq('date', record['date']).execute()
            if existing.data:
                return jsonify({"success": False, "error": f"Attendance already marked for student {record['student_id']} on {record['date']}"}), 400
        
        # Insert all records
        for record in attendance_records:
            attendance_data = {
                'student_id': record['student_id'],
                'subject_id': record['subject_id'],
                'faculty_id': record['faculty_id'],
                'date': record['date'],
                'status': record['status'],
                'remarks': record.get('remarks'),
                'created_at': datetime.now().isoformat()
            }
            supabase.table('attendance').insert(attendance_data).execute()
        
        return jsonify({"success": True, "message": f"Attendance marked for {len(attendance_records)} students"}), 201
    
    else:
        # Single attendance marking
        required_fields = ['student_id', 'subject_id', 'faculty_id', 'date', 'status']
        
        for field in required_fields:
            if field not in data:
                return jsonify({"success": False, "error": f"Missing required field: {field}"}), 400
        
        # Validate entities exist
        student_result = supabase.table('students').select('id').eq('id', data['student_id']).execute()
        if not student_result.data:
            return jsonify({"success": False, "error": "Invalid student_id"}), 400
        
        subject_result = supabase.table('subjects').select('id').eq('id', data['subject_id']).execute()
        if not subject_result.data:
            return jsonify({"success": False, "error": "Invalid subject_id"}), 400
        
        faculty_result = supabase.table('faculty').select('id').eq('id', data['faculty_id']).execute()
        if not faculty_result.data:
            return jsonify({"success": False, "error": "Invalid faculty_id"}), 400
        
        # Validate status
        valid_statuses = ['present', 'absent', 'late']
        if data['status'] not in valid_statuses:
            return jsonify({"success": False, "error": f"Invalid status. Must be one of: {valid_statuses}"}), 400
        
        # Check if attendance already exists
        existing = supabase.table('attendance').select('id').eq('student_id', data['student_id']).eq('subject_id', data['subject_id']).eq('date', data['date']).execute()
        if existing.data:
            return jsonify({"success": False, "error": "Attendance already marked for this student on this date"}), 400
        
        attendance_data = {
            'student_id': data['student_id'],
            'subject_id': data['subject_id'],
            'faculty_id': data['faculty_id'],
            'date': data['date'],
            'status': data['status'],
            'remarks': data.get('remarks'),
            'created_at': datetime.now().isoformat()
        }
        
        result = supabase.table('attendance').insert(attendance_data).execute()
        return jsonify({"success": True, "data": result.data[0] if result.data else {}}), 201

@attendance_bp.route('/attendance/<int:attendance_id>', methods=['PUT'])
@handle_errors
def update_attendance(attendance_id):
    """Update attendance record"""
    data = request.get_json()
    
    # Validate status if provided
    if 'status' in data:
        valid_statuses = ['present', 'absent', 'late']
        if data['status'] not in valid_statuses:
            return jsonify({"success": False, "error": f"Invalid status. Must be one of: {valid_statuses}"}), 400
    
    result = supabase.table('attendance').update(data).eq('id', attendance_id).execute()
    if not result.data:
        return jsonify({"success": False, "error": "Attendance record not found"}), 404
    
    return jsonify({"success": True, "data": result.data[0]})

@attendance_bp.route('/attendance/<int:attendance_id>', methods=['DELETE'])
@handle_errors
def delete_attendance(attendance_id):
    """Delete attendance record"""
    result = supabase.table('attendance').delete().eq('id', attendance_id).execute()
    if not result.data:
        return jsonify({"success": False, "error": "Attendance record not found"}), 404
    
    return jsonify({"success": True, "message": "Attendance record deleted successfully"})

# Attendance Analytics
@attendance_bp.route('/attendance/analytics/student/<int:student_id>', methods=['GET'])
@handle_errors
def get_student_attendance_analytics(student_id):
    """Get attendance analytics for a specific student"""
    # Get all attendance records for the student
    result = supabase.table('attendance').select('''
        *,
        subjects (
            name,
            code
        )
    ''').eq('student_id', student_id).execute()
    
    if not result.data:
        return jsonify({"success": False, "error": "No attendance records found for this student"}), 404
    
    attendance_data = result.data
    total_classes = len(attendance_data)
    
    # Calculate overall statistics
    present_count = len([record for record in attendance_data if record['status'] == 'present'])
    absent_count = len([record for record in attendance_data if record['status'] == 'absent'])
    late_count = len([record for record in attendance_data if record['status'] == 'late'])
    
    overall_percentage = (present_count / total_classes) * 100 if total_classes > 0 else 0
    
    # Subject-wise analytics
    subject_analytics = {}
    for record in attendance_data:
        subject_name = record['subjects']['name']
        if subject_name not in subject_analytics:
            subject_analytics[subject_name] = {
                'subject_code': record['subjects']['code'],
                'total_classes': 0,
                'present': 0,
                'absent': 0,
                'late': 0
            }
        
        subject_analytics[subject_name]['total_classes'] += 1
        subject_analytics[subject_name][record['status']] += 1
    
    # Calculate percentage for each subject
    for subject in subject_analytics:
        total = subject_analytics[subject]['total_classes']
        present = subject_analytics[subject]['present']
        subject_analytics[subject]['percentage'] = (present / total) * 100 if total > 0 else 0
    
    analytics = {
        'student_id': student_id,
        'overall_statistics': {
            'total_classes': total_classes,
            'present_count': present_count,
            'absent_count': absent_count,
            'late_count': late_count,
            'attendance_percentage': round(overall_percentage, 2)
        },
        'subject_wise_analytics': subject_analytics
    }
    
    return jsonify({"success": True, "data": analytics})

@attendance_bp.route('/attendance/analytics/subject/<int:subject_id>', methods=['GET'])
@handle_errors
def get_subject_attendance_analytics(subject_id):
    """Get attendance analytics for a specific subject"""
    # Get all attendance records for the subject
    result = supabase.table('attendance').select('''
        *,
        students (
            register_number,
            full_name
        )
    ''').eq('subject_id', subject_id).execute()

    if not result.data:
        return jsonify({"success": False, "error": "No attendance records found for this subject"}), 404

    attendance_data = result.data
    total_records = len(attendance_data)

    # Calculate overall statistics
    present_count = len([record for record in attendance_data if record['status'] == 'present'])
    absent_count = len([record for record in attendance_data if record['status'] == 'absent'])
    late_count = len([record for record in attendance_data if record['status'] == 'late'])

    overall_percentage = (present_count / total_records) * 100 if total_records > 0 else 0

    # Student-wise analytics
    student_analytics = {}
    for record in attendance_data:
        student_id = record['student_id']
        if student_id not in student_analytics:
            student_analytics[student_id] = {
                'student_name': record['students']['full_name'],
                'register_number': record['students']['register_number'],
                'total_classes': 0,
                'present': 0,
                'absent': 0,
                'late': 0
            }

        student_analytics[student_id]['total_classes'] += 1
        student_analytics[student_id][record['status']] += 1

    # Calculate percentage for each student and identify defaulters
    defaulters = []
    for student_id, data in student_analytics.items():
        total = data['total_classes']
        present = data['present']
        percentage = (present / total) * 100 if total > 0 else 0
        data['percentage'] = round(percentage, 2)

        # Consider students with less than 75% attendance as defaulters
        if percentage < 75:
            defaulters.append({
                'student_id': student_id,
                'student_name': data['student_name'],
                'register_number': data['register_number'],
                'attendance_percentage': percentage,
                'absent_classes': data['absent']
            })

    analytics = {
        'subject_id': subject_id,
        'overall_statistics': {
            'total_records': total_records,
            'present_count': present_count,
            'absent_count': absent_count,
            'late_count': late_count,
            'overall_percentage': round(overall_percentage, 2)
        },
        'student_wise_analytics': student_analytics,
        'defaulters': defaulters,
        'defaulters_count': len(defaulters)
    }

    return jsonify({"success": True, "data": analytics})

@attendance_bp.route('/attendance/analytics/faculty/<int:faculty_id>', methods=['GET'])
@handle_errors
def get_faculty_attendance_analytics(faculty_id):
    """Get attendance analytics for classes taken by a specific faculty"""
    # Get all attendance records for the faculty
    result = supabase.table('attendance').select('''
        *,
        subjects (
            name,
            code
        ),
        students (
            register_number,
            full_name
        )
    ''').eq('faculty_id', faculty_id).execute()

    if not result.data:
        return jsonify({"success": False, "error": "No attendance records found for this faculty"}), 404

    attendance_data = result.data
    total_records = len(attendance_data)

    # Calculate overall statistics
    present_count = len([record for record in attendance_data if record['status'] == 'present'])
    absent_count = len([record for record in attendance_data if record['status'] == 'absent'])
    late_count = len([record for record in attendance_data if record['status'] == 'late'])

    overall_percentage = (present_count / total_records) * 100 if total_records > 0 else 0

    # Subject-wise analytics
    subject_analytics = {}
    for record in attendance_data:
        subject_name = record['subjects']['name']
        if subject_name not in subject_analytics:
            subject_analytics[subject_name] = {
                'subject_code': record['subjects']['code'],
                'total_classes': 0,
                'present': 0,
                'absent': 0,
                'late': 0
            }

        subject_analytics[subject_name]['total_classes'] += 1
        subject_analytics[subject_name][record['status']] += 1

    # Calculate percentage for each subject
    for subject in subject_analytics:
        total = subject_analytics[subject]['total_classes']
        present = subject_analytics[subject]['present']
        subject_analytics[subject]['percentage'] = round((present / total) * 100, 2) if total > 0 else 0

    analytics = {
        'faculty_id': faculty_id,
        'overall_statistics': {
            'total_records': total_records,
            'present_count': present_count,
            'absent_count': absent_count,
            'late_count': late_count,
            'overall_percentage': round(overall_percentage, 2)
        },
        'subject_wise_analytics': subject_analytics
    }

    return jsonify({"success": True, "data": analytics})

@attendance_bp.route('/attendance/defaulters', methods=['GET'])
@handle_errors
def get_attendance_defaulters():
    """Get list of students with low attendance across all subjects"""
    # Get all attendance records
    result = supabase.table('attendance').select('''
        *,
        students (
            register_number,
            full_name,
            courses (
                name,
                code
            )
        ),
        subjects (
            name,
            code
        )
    ''').execute()

    if not result.data:
        return jsonify({"success": True, "data": []}), 200

    attendance_data = result.data

    # Calculate student-wise attendance
    student_analytics = {}
    for record in attendance_data:
        student_id = record['student_id']
        if student_id not in student_analytics:
            student_analytics[student_id] = {
                'student_name': record['students']['full_name'],
                'register_number': record['students']['register_number'],
                'course': record['students']['courses']['name'] if record['students']['courses'] else 'N/A',
                'total_classes': 0,
                'present': 0,
                'absent': 0,
                'late': 0,
                'subjects': set()
            }

        student_analytics[student_id]['total_classes'] += 1
        student_analytics[student_id][record['status']] += 1
        student_analytics[student_id]['subjects'].add(record['subjects']['name'])

    # Identify defaulters (students with less than 75% attendance)
    defaulters = []
    threshold = float(request.args.get('threshold', 75))  # Default threshold is 75%

    for student_id, data in student_analytics.items():
        total = data['total_classes']
        present = data['present']
        percentage = (present / total) * 100 if total > 0 else 0

        if percentage < threshold:
            defaulters.append({
                'student_id': student_id,
                'student_name': data['student_name'],
                'register_number': data['register_number'],
                'course': data['course'],
                'total_classes': total,
                'present_classes': present,
                'absent_classes': data['absent'],
                'late_classes': data['late'],
                'attendance_percentage': round(percentage, 2),
                'subjects_count': len(data['subjects'])
            })

    # Sort by attendance percentage (lowest first)
    defaulters.sort(key=lambda x: x['attendance_percentage'])

    return jsonify({
        "success": True,
        "data": defaulters,
        "count": len(defaulters),
        "threshold": threshold
    })

File: Updated_CMS-main/backend/routes/attendance.js
const express = require('express');
const router = express.Router();
const { supabase } = require('../supabase_client');

// Get all attendance records with filters
router.get('/', async (req, res) => {
  try {
    const { 
      student_id, 
      subject_id, 
      date, 
      status, 
      start_date, 
      end_date,
      page = 1,
      limit = 20
    } = req.query;
    
    const offset = (page - 1) * limit;
    
    let query = supabase
      .from('attendance')
      .select('*, students(*), subjects(*)', { count: 'exact' });

    // Apply filters
    if (student_id) {
      query = query.eq('student_id', student_id);
    }
    
    if (subject_id) {
      query = query.eq('subject_id', subject_id);
    }
    
    if (date) {
      query = query.eq('date', date);
    }
    
    if (status) {
      query = query.eq('status', status);
    }
    
    if (start_date && end_date) {
      query = query.gte('date', start_date).lte('date', end_date);
    }
    
    // Apply pagination
    query = query.range(offset, offset + limit - 1);
    
    // Order by date descending
    query = query.order('date', { ascending: false });
    
    const { data, error, count } = await query;

    if (error) throw error;

    res.json({
      status: 'success',
      data,
      pagination: {
        total: count,
        page: parseInt(page),
        limit: parseInt(limit),
        totalPages: Math.ceil(count / limit)
      }
    });
  } catch (error) {
    console.error('Error fetching attendance records:', error);
    res.status(500).json({ status: 'error', message: 'Failed to fetch attendance records' });
  }
});

// Get single attendance record by ID
router.get('/:id', async (req, res) => {
  try {
    const { id } = req.params;
    
    const { data, error } = await supabase
      .from('attendance')
      .select('*, students(*), subjects(*)')
      .eq('id', id)
      .single();

    if (error) throw error;
    if (!data) {
      return res.status(404).json({ status: 'error', message: 'Attendance record not found' });
    }

    res.json({ status: 'success', data });
  } catch (error) {
    console.error('Error fetching attendance record:', error);
    res.status(500).json({ status: 'error', message: 'Failed to fetch attendance record' });
  }
});

// Create new attendance record
router.post('/', async (req, res) => {
  try {
    const { student_id, subject_id, date, status, remarks } = req.body;
    
    // Validate required fields
    if (!student_id || !subject_id || !date || !status) {
      return res.status(400).json({
        status: 'error',
        message: 'Student ID, subject ID, date, and status are required'
      });
    }
    
    // Validate status
    if (!['present', 'absent', 'late', 'excused'].includes(status)) {
      return res.status(400).json({
        status: 'error',
        message: 'Invalid status. Must be one of: present, absent, late, excused'
      });
    }
    
    // Check if student exists
    const { data: student, error: studentError } = await supabase
      .from('students')
      .select('id')
      .eq('id', student_id)
      .single();

    if (studentError || !student) {
      return res.status(400).json({ 
        status: 'error', 
        message: 'Invalid student ID' 
      });
    }
    
    // Check if subject exists
    const { data: subject, error: subjectError } = await supabase
      .from('subjects')
      .select('id')
      .eq('id', subject_id)
      .single();

    if (subjectError || !subject) {
      return res.status(400).json({ 
        status: 'error', 
        message: 'Invalid subject ID' 
      });
    }
    
    // Check if attendance record already exists for this student, subject, and date
    const { data: existingRecord, error: existingError } = await supabase
      .from('attendance')
      .select('id')
      .eq('student_id', student_id)
      .eq('subject_id', subject_id)
      .eq('date', date)
      .maybeSingle();

    if (existingError) throw existingError;
    
    if (existingRecord) {
      return res.status(400).json({
        status: 'error',
        message: 'Attendance record already exists for this student, subject, and date'
      });
    }
    
    // Create new attendance record
    const { data, error } = await supabase
      .from('attendance')
      .insert([{
        student_id,
        subject_id,
        date,
        status,
        remarks: remarks || null
      }])
      .select()
      .single();

    if (error) throw error;

    res.status(201).json({ status: 'success', data });
  } catch (error) {
    console.error('Error creating attendance record:', error);
    res.status(500).json({ status: 'error', message: 'Failed to create attendance record' });
  }
});

// Update attendance record
router.put('/:id', async (req, res) => {
  try {
    const { id } = req.params;
    const { status, remarks } = req.body;
    
    // Validate status if provided
    if (status && !['present', 'absent', 'late', 'excused'].includes(status)) {
      return res.status(400).json({
        status: 'error',
        message: 'Invalid status. Must be one of: present, absent, late, excused'
      });
    }
    
    // Build update object
    const updates = {};
    if (status) updates.status = status;
    if (remarks !== undefined) updates.remarks = remarks;
    
    // Add updated_at timestamp
    updates.updated_at = new Date().toISOString();
    
    const { data, error } = await supabase
      .from('attendance')
      .update(updates)
      .eq('id', id)
      .select()
      .single();

    if (error) throw error;
    if (!data) {
      return res.status(404).json({ status: 'error', message: 'Attendance record not found' });
    }

    res.json({ status: 'success', data });
  } catch (error) {
    console.error('Error updating attendance record:', error);
    res.status(500).json({ status: 'error', message: 'Failed to update attendance record' });
  }
});

// Delete attendance record
router.delete('/:id', async (req, res) => {
  try {
    const { id } = req.params;
    
    const { error } = await supabase
      .from('attendance')
      .delete()
      .eq('id', id);

    if (error) throw error;

    res.json({ status: 'success', message: 'Attendance record deleted successfully' });
  } catch (error) {
    console.error('Error deleting attendance record:', error);
    res.status(500).json({ status: 'error', message: 'Failed to delete attendance record' });
  }
});

// Bulk upload attendance
router.post('/bulk', async (req, res) => {
  try {
    const { records } = req.body;
    
    if (!Array.isArray(records) || records.length === 0) {
      return res.status(400).json({
        status: 'error',
        message: 'No attendance records provided'
      });
    }
    
    // Validate each record
    const validStatuses = ['present', 'absent', 'late', 'excused'];
    
    for (const record of records) {
      if (!record.student_id || !record.subject_id || !record.date || !record.status) {
        return res.status(400).json({
          status: 'error',
          message: 'Each record must have student_id, subject_id, date, and status'
        });
      }
      
      if (!validStatuses.includes(record.status)) {
        return res.status(400).json({
          status: 'error',
          message: `Invalid status '${record.status}'. Must be one of: ${validStatuses.join(', ')}`
        });
      }
      
      // Check if student exists
      const { data: student, error: studentError } = await supabase
        .from('students')
        .select('id')
        .eq('id', record.student_id)
        .single();

      if (studentError || !student) {
        return res.status(400).json({ 
          status: 'error', 
          message: `Invalid student ID: ${record.student_id}`
        });
      }
      
      // Check if subject exists
      const { data: subject, error: subjectError } = await supabase
        .from('subjects')
        .select('id')
        .eq('id', record.subject_id)
        .single();

      if (subjectError || !subject) {
        return res.status(400).json({ 
          status: 'error', 
          message: `Invalid subject ID: ${record.subject_id}`
        });
      }
    }
    
    // Check for existing records to avoid duplicates
    const existingRecords = [];
    
    for (const record of records) {
      const { data, error } = await supabase
        .from('attendance')
        .select('id')
        .eq('student_id', record.student_id)
        .eq('subject_id', record.subject_id)
        .eq('date', record.date)
        .maybeSingle();
        
      if (error) throw error;
      
      if (data) {
        existingRecords.push({
          student_id: record.student_id,
          subject_id: record.subject_id,
          date: record.date
        });
      }
    }
    
    if (existingRecords.length > 0) {
      return res.status(400).json({
        status: 'error',
        message: 'Some attendance records already exist',
        existing_records: existingRecords
      });
    }
    
    // Prepare records for insertion
    const recordsToInsert = records.map(record => ({
      student_id: record.student_id,
      subject_id: record.subject_id,
      date: record.date,
      status: record.status,
      remarks: record.remarks || null
    }));
    
    // Insert records in batches of 100 (Supabase limit)
    const BATCH_SIZE = 100;
    const results = [];
    
    for (let i = 0; i < recordsToInsert.length; i += BATCH_SIZE) {
      const batch = recordsToInsert.slice(i, i + BATCH_SIZE);
      const { data, error } = await supabase
        .from('attendance')
        .insert(batch)
        .select();
        
      if (error) throw error;
      
      results.push(...data);
    }
    
    res.status(201).json({
      status: 'success',
      data: results,
      count: results.length
    });
  } catch (error) {
    console.error('Error in bulk attendance upload:', error);
    res.status(500).json({ 
      status: 'error', 
      message: 'Failed to process bulk attendance upload',
      error: error.message
    });
  }
});

// Get attendance summary for a student
router.get('/summary/student/:student_id', async (req, res) => {
  try {
    const { student_id } = req.params;
    const { subject_id, start_date, end_date } = req.query;
    
    // Check if student exists
    const { data: student, error: studentError } = await supabase
      .from('students')
      .select('id, course_id')
      .eq('id', student_id)
      .single();

    if (studentError || !student) {
      return res.status(404).json({ 
        status: 'error', 
        message: 'Student not found' 
      });
    }
    
    // Build query
    let query = supabase
      .from('attendance')
      .select('*, subjects(*, courses(*))')
      .eq('student_id', student_id);
    
    // Apply filters
    if (subject_id) {
      query = query.eq('subject_id', subject_id);
    }
    
    if (start_date && end_date) {
      query = query.gte('date', start_date).lte('date', end_date);
    }
    
    const { data: attendance, error } = await query;
    
    if (error) throw error;
    
    // Calculate summary
    const summary = {};
    
    attendance.forEach(record => {
      const subjectId = record.subject_id;
      
      if (!summary[subjectId]) {
        summary[subjectId] = {
          subject: record.subjects,
          total: 0,
          present: 0,
          absent: 0,
          late: 0,
          excused: 0,
          percentage: 0
        };
      }
      
      summary[subjectId].total++;
      
      switch (record.status) {
        case 'present':
          summary[subjectId].present++;
          break;
        case 'absent':
          summary[subjectId].absent++;
          break;
        case 'late':
          summary[subjectId].late++;
          break;
        case 'excused':
          summary[subjectId].excused++;
          break;
      }
      
      // Calculate percentage (excluding excused absences)
      const totalCounted = summary[subjectId].total - summary[subjectId].excused;
      const presentCount = summary[subjectId].present + summary[subjectId].late;
      
      summary[subjectId].percentage = totalCounted > 0 
        ? Math.round((presentCount / totalCounted) * 100) 
        : 0;
    });
    
    // Convert to array
    const result = Object.values(summary);
    
    res.json({
      status: 'success',
      data: result,
      student: {
        id: student.id,
        name: student.first_name ? `${student.first_name} ${student.last_name || ''}`.trim() : null
      }
    });
  } catch (error) {
    console.error('Error fetching attendance summary:', error);
    res.status(500).json({ 
      status: 'error', 
      message: 'Failed to fetch attendance summary' 
    });
  }
});

// Get attendance summary for a subject
router.get('/summary/subject/:subject_id', async (req, res) => {
  try {
    const { subject_id } = req.params;
    const { date, start_date, end_date } = req.query;
    
    // Check if subject exists
    const { data: subject, error: subjectError } = await supabase
      .from('subjects')
      .select('id, name, code, courses(*)')
      .eq('id', subject_id)
      .single();

    if (subjectError || !subject) {
      return res.status(404).json({ 
        status: 'error', 
        message: 'Subject not found' 
      });
    }
    
    // Build query
    let query = supabase
      .from('attendance')
      .select('*, students(*)')
      .eq('subject_id', subject_id);
    
    // Apply date filters
    if (date) {
      query = query.eq('date', date);
    } else if (start_date && end_date) {
      query = query.gte('date', start_date).lte('date', end_date);
    }
    
    const { data: attendance, error } = await query;
    
    if (error) throw error;
    
    // Group by student
    const studentSummary = {};
    
    attendance.forEach(record => {
      const studentId = record.student_id;
      
      if (!studentSummary[studentId]) {
        studentSummary[studentId] = {
          student: record.students,
          total: 0,
          present: 0,
          absent: 0,
          late: 0,
          excused: 0,
          percentage: 0
        };
      }
      
      studentSummary[studentId].total++;
      
      switch (record.status) {
        case 'present':
          studentSummary[studentId].present++;
          break;
        case 'absent':
          studentSummary[studentId].absent++;
          break;
        case 'late':
          studentSummary[studentId].late++;
          break;
        case 'excused':
          studentSummary[studentId].excused++;
          break;
      }
      
      // Calculate percentage (excluding excused absences)
      const student = studentSummary[studentId];
      const totalCounted = student.total - student.excused;
      const presentCount = student.present + student.late;
      
      student.percentage = totalCounted > 0 
        ? Math.round((presentCount / totalCounted) * 100) 
        : 0;
    });
    
    // Calculate overall summary
    const overall = {
      total_students: Object.keys(studentSummary).length,
      total_records: attendance.length,
      present: 0,
      absent: 0,
      late: 0,
      excused: 0,
      average_percentage: 0
    };
    
    let totalPercentage = 0;
    let studentCount = 0;
    
    Object.values(studentSummary).forEach(student => {
      overall.present += student.present;
      overall.absent += student.absent;
      overall.late += student.late;
      overall.excused += student.excused;
      
      if (student.total > student.excused) {
        totalPercentage += student.percentage;
        studentCount++;
      }
    });
    
    overall.average_percentage = studentCount > 0 
      ? Math.round(totalPercentage / studentCount) 
      : 0;
    
    // Convert to array
    const result = {
      subject,
      overall,
      students: Object.values(studentSummary)
    };
    
    res.json({
      status: 'success',
      data: result
    });
  } catch (error) {
    console.error('Error fetching subject attendance summary:', error);
    res.status(500).json({ 
      status: 'error', 
      message: 'Failed to fetch subject attendance summary' 
    });
  }
});

module.exports = router;
